### ISSUE TITLE
How exactly does a wsiget use ui_base?

### ISSUE THREAD

**Issue by AllanOricil** [FlowFuse/node-red-dashboard#1289]
I read the dashboard 2 docs and understood that widgets are node-red nodes, and that they "extend" the ui_base config node, but I could not understand how exactly this happens. Somehow the widget's node has some additional methods called register/deregister. Probably has other, but I did not read everything yet. So, when/how exactly does a widget receive those 2 methods?

**Comment by joepavitt** [comment]
The UI Base is the single parent object for a Dashboard. 

In the constructor for each Widget, they get reference to the UI Base (normally via their assigned group/page). They call base.register, and pass themselves into thag function call

**Comment by AllanOricil** [comment]
@joepavitt what do you think of my idea of making ui_base into a parent class called WidgetNode, and then widgets would extends this class?  I think it is possible. Could you help me to see problems I would have trying to do it?

**Comment by joepavitt** [comment]
The UI Base has settings and configuration of its own. The widgets aren't necessarily extensions in a Class sense. They are different entities. There is only ever one Base. Having a Node an extension of Base would mean you suddenly have 100's of Dashboards/Based

If we consider Class hierarchy, there could be benefit in having a common `WidgetNode` class that the nodes, in order to share feature sets, etc.

The `base.register` function is the Widget's way of communicating to the Dashboard to say "Hey, I exist"

**Comment by joepavitt** [comment]
My use of the word "extends" here is likely misleading. 

From a Dashboard Hierarchy we have:

- Base
    - Theme 
    - Page
        - Group
            - Widget

That doesn't mean that there is Class Inheritance here though. Fundamentally, these _are_ the Classes, then each node in a flow (or config node) in an instance of the respective class

**Comment by joepavitt** [comment]
Also worth noting, and I may be remembering this wrong, the nodes themselves dont need a register/deregister function.

The `base.register()` is just called as part of the Node's constructor 

Maybe they do need a deregister function?

**Comment by AllanOricil** [comment]
@joepavitt  

> Maybe they do need a deregister function? I can't quite remember how I handle that, as I'm not at a keyboard

I believe that if a node can call `this.deregister()`, maybe there are situations where the node would also call `this.register()` afterward. Maybe to do a "hard reset" somewhere? Do you know if there is any Widget that does it?



> The `base.register()` is just called as part of the Node's constructor

I can ensure this is called inside the parent's constructor, or in the mixin. This way devs won't need to manually write it. The same way I did for `RED.nodes.registerType` or `RED.nodes.createNode` 

https://github.com/AllanOricil/nrg/blob/main/templates/server/entrypoint.handlebars
https://github.com/AllanOricil/node-red-node

**Comment by AllanOricil** [comment]
@joepavitt 
What do you think of this model? 

```js
export default class WidgetNode extends Node {

    #scope: Base | Theme | Page | Group;  // each instance of a widget can have its own location

    constructor(config){
        super(config);
        //initialize specific WidgetNode properties, such as state store
    }

    // specific WidgetNode methods, like register/ deregister
    
    register(){
         this.#scope.register(this);
    }
    
    deregister(){
         this.#scope.deregister(this);
    }
}

```

**Comment by joepavitt** [comment]
That could work, bare in mind though some widgets are permitted to be multiple, e.g. `ui-template` can be switched between group, page and UI

**Comment by AllanOricil** [comment]
> That could work, bare in mind though some widgets are permitted to be multiple, e.g. `ui-template` can be switched between group, page and UI

How does a Widget switch between "ui scopes" (have to find a better name)?

**Comment by colinl** [comment]
It is defined in template node's configuration


![image](https://github.com/user-attachments/assets/b15d88fb-c874-4f71-96f7-85ab28260a61)

**Comment by AllanOricil** [comment]
> It is defined in template node's configuration
> 
> 
> ![image](https://github.com/user-attachments/assets/b15d88fb-c874-4f71-96f7-85ab28260a61)
> 

@colinl every widget has those fields shown in that image? How can I get to that configuration form?

**Comment by AllanOricil** [comment]
I started to have a better understanding after reading the config nodes

<img width="177" alt="image" src="https://github.com/user-attachments/assets/861e03dc-2cce-459f-8c3e-4e661ea96992">

I think I can start modeling it into classes. I will try to do the base ones, and a button.

**Comment by AllanOricil** [comment]
One last question.

@colinl @joepavitt 

What are all possible event names people can use?

```js
const evts = {
            beforeSend: async function (msg)
}

group.register(node, config, evts)
```

```js
export default Widget exnteds UIEntity {
   
     /*
     * It can be implemented by a child class
     * @abstract
     */
     static onBeforeSend(msg){}
}
```

```js
export default MyWidget extends Widget {

      constructor(config){
      }
      
      static onBeforeSend(msg) {
            ....
      }
      
      // all other event handlers
}
```


The inside the Mixin, I will do some sorcery js to make sure these "on" events are registered.

**Comment by AllanOricil** [comment]
@joepavitt @colinl could you help with the last comment?

**Comment by colinl** [comment]
Not me, I don't know the answer.

**Comment by joepavitt** [comment]
@AllanOricil https://dashboard.flowfuse.com/contributing/guides/registration.html#evts

**Comment by joepavitt** [comment]
>  every widget has those fields shown in that image? How can I get to that configuration form?

Drop a Dashboard "template" node into the Editor, then double click it to configure it. Only the "Template" node as the option to switch between "parent scope" like this.

**Comment by AllanOricil** [comment]
> @AllanOricil https://dashboard.flowfuse.com/contributing/guides/registration.html#evts

Is the `onInput` from evts the same as the node's on input `this.on("input", onInputHandler())`?

**Comment by joepavitt** [comment]
Not quite, the one you've quoted there is the _client-side_ handler, the `onInput` I linked to is the server-side handler

**Comment by AllanOricil** [comment]
> Not quite, the one you've quoted there is the _client-side_ handler (Vue Widget), the `onInput` I linked to is the server-side handler (the Node-RED node)



So all `evts` handlers are sent to the client? I thought they executed on the server, because they are declared in the server js of a node.

**Comment by joepavitt** [comment]
No, they're separate. What a node does on input server-side (defined in the .js` file) will differ the the client-side behaviour of a widget when receiving a message (defined in the `vue` file)

**Comment by joepavitt** [comment]
Sorry, I thought `this.on("input", onInputHandler())` had been taken from a `vue` file as it's _very_ similar to our code there too.

Yes, if you checkout the `node.register` function in `ui-base.js`, you'll see that the `on('input')` handler is setup for the relevant node and the `evts.onInput` is passed there (if provided)

**Comment by AllanOricil** [comment]
> Sorry, I thought `this.on("input", onInputHandler())` had been taken from a `vue` file as it's _very_ similar to our code there too.
> 
> Yes, if you checkout the `node.register` function in `ui-base.js`, you'll see that the `on('input')` handler is setup for the relevant node and the `evts.onInput` is passed there (if provided)

I would need a new name for that handler to put it on the class alongside the node's this.on("input")

I can't have 2 `onInput` methods in the class.  What about `onData` for the `evts.onInput`?

**Comment by joepavitt** [comment]
not sure why the class needs that definition though? It needs `evts`, which is then an object mapped to x functions/objets/booleans

**Comment by AllanOricil** [comment]
> not sure why the class needs that definition though? It needs `evts`, which is then an object mapped to x functions/objets/booleans

When one of those handlers is triggered, can't they access "instance" data? I can have multiple widget nodes, all with the same handler, but each with its own instance scope. Those events are part of the Widget, so I'm trying to encapsulate them normalized as single functions, instead of using an evts attribute. It looks cleaner in my opinion.
